includes/class-wc-gateway-amazon-payments-advanced.php

```diff
@@ -2537,4 +2598,88 @@ public function load_scripts_globally_if_button_enabled_on_cart( $load_scripts )
 		return $load_scripts ? $load_scripts : ( ! empty( $this->settings['mini_cart_button'] ) && 'yes' === $this->settings['mini_cart_button'] );
 	}
 
+	/**
+	 * Loads scripts on product pages when the setting to display Amazon Pay button on products is enabled.
+	 *
+	 * @param bool $load_scripts
+	 * @return bool
+	 */
+	public function load_scripts_on_product_pages( $load_scripts ) {
+		return $load_scripts ? $load_scripts : is_product();
+	}
+
+	/**
+	 * Restores a cart stored in session.
+	 *
+	 * @return void
+	 */
+	public function remember_forgotten_cart() {
+		$forgotten_cart = WC()->session->get( 'amazon_pay_forgotten_cart' );
+		if ( $forgotten_cart ) {
+			foreach ( $forgotten_cart as $forgotten_item ) {
+				if ( $forgotten_item['product_variation_id'] ) {
+					$variation = new WC_Product_Variation( $forgotten_item['product_variation_id'] );
+					$attrs     = $variation->get_variation_attributes();
+				} else {
+					$attrs = array();
+				}
+				WC()->cart->add_to_cart( $forgotten_item['product_id'], $forgotten_item['quantity'], $forgotten_item['product_variation_id'], $attrs );
+			}
+			WC()->session->set( 'amazon_pay_forgotten_cart', null );
+		}
+	}
+
+	/**
+	 * Stores current cart in session and creates a new one.
+	 * Notifies JS of the current checkout session config params.
+	 *
+	 * Ajax endpoint hooked in the 'apa_change_wc_cart' action.
+	 *
+	 * @return void
+	 */
+	public function ajax_apa_change_wc_carts() {
```

Comment:
IMO this needs to go low level when saving the contents of the cart as there may be other data there added by 3rd party plugins.

Reply:
I think we should be able to save the whole dataset returned by `get_cart_contents() ` and restore it using `set_cart_contents()`

Reply:
Went deeper into woo's code and found out how to restore a cart using woo's methods.